### PR TITLE
feat: Make the Query type conform to Identifiable

### PR DESCRIPTION
**Pull request by vdkdamian**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/netreconlab/Parse-Swift/security/policy).
- [x] I am creating this PR in reference to an [issue](https://github.com/netreconlab/Parse-Swift/issues/41).

### Issue Description
Make Queries `Identifiable`

### Approach
Added the protocol

### TODOs before merging

- [x] Add tests
- [ ] Add entry to changelog
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)

### CODE REVIEW

**Comment by vdkdamian** [comment]
@cbaker6 Would you allow this change? Then I can work on the tests

**Comment by cbaker6** [comment]
I'm not sure, what's the reason(s) for needing a query to be `Identifiable`?

**Comment by vdkdamian** [comment]
> I'm not sure, what's the reason(s) for needing a query to be `Identifiable`?

I need it, cause I am storing objects based on the query. I could just use the Fork on my side, but I would prefer to be able to use this version without the need to follow it up.

Also think it doesn't hurt to make it `Identifiable`. Maybe some other people can make use of it. 
Let me know what you think!

**Comment by cbaker6** [comment]
> I need it, cause I am storing objects based on the query. I could just use the Fork on my side, but I would prefer to be able to use this version without the need to follow it up.

I don't see a reason for needing to store the query. If you really need it, you shouldn't need a fork, you should just be able to add your extension to your app(s). Depending on your use cases, you can even create a framework that depends on parse-swift with all of your extensions if you think they will be useful to others.

> Also think it doesn't hurt to make it Identifiable. Maybe some other people can make use of it.
Let me know what you think!

I'm also not sure if this is the best way to make a `Query` identifiable as it takes some computation each time to compare ID's. In addition, adding this means it should be maintained.

**Comment by vdkdamian** [comment]
> I don't see a reason for needing to store the query. If you really need it, you shouldn't need a fork, you should just be able to add your extension to your app(s).

As mentioned, I am unable to access the internal variables. So creating an id of it is not possible via an extension...

**Comment by cbaker6** [comment]
> As mentioned, I am unable to access the internal variables. So creating an id of it is not possible via an extension...

Query is `Codable` and `Equatable`. I currently don't see a compelling reason to make it `Identifiable` as a Query can be stored in it's JSON form along with being compared to other queries. Depending on the variables, I might be willing to make them public. Which variables would need to be changed and what would be your suggested changes to them? Note that frameworks are usually cautious of making internal and private properties public, so making these type of changes need to be accompanied with convincing arguments.

**Comment by vdkdamian** [comment]
> Query is `Codable` and `Equatable`. I currently don't see a compelling reason to make it `Identifiable` as a Query can be stored in it's JSON form along with being compared to other queries. 

That's a good point. But for 4 variables, you are using `sets`. 
Which results in a random collection where the values can switch position. 
This indirectly means that the id I need can change. Which is not something I need, nor which makes it `Identifiable`

> Depending on the variables, I might be willing to make them public. Which variables would need to be changed and what would be your suggested changes to them? 

It's the 4 variables that are sets.

> Note that frameworks are usually cautious of making internal and private properties public, so making these type of changes need to be accompanied with convincing arguments.

I am aware of this. That was the main reason I suggested to make it `Identifiable`. Another option that can be considered is to make a public variable where the code from the Id is returned. Then we can skip the `Identifiable` protocol.

I don't know which solution you prefer.

**Comment by cbaker6** [comment]
> Which results in a random collection where the values can switch position.

The query is equatable, this shouldn’t matter. Are you saying 2 queries with the same sets don’t return equal?

> I don't know which solution you prefer.

I’m still not convinced a query needs to be identifiable or persisted. Even if someone needs to store it, it can be stored in JSON form and decoded to check for equality. It sounds like you would like to treat a query as a ParseObject, in which it is not. Making the properties public opens up the possibility for headaches.

File: Tests/ParseSwiftTests/ParseQueryTests.swift [hunk 1]
```diff
@@ -104,6 +104,22 @@ class ParseQueryTests: XCTestCase { // swiftlint:disable:this type_body_length
         XCTAssertEqual(query5.`where`.constraints.values.count, 0)
     }
 
+    func testIdentifier() {
```
Comment (cbaker6) [inline Tests/ParseSwiftTests/ParseQueryTests.swift:107]: More appropriate tests are needed to test if your implementation of Identifiable is correct. For example, from your design, 2 different queries with the same properties for the same ParseObject should produce the exact same ID. Conversely,  2 different queries with slightly different properties for the same ParseObject should produce a different ID

File: Sources/ParseSwift/Types/Query.swift [hunk 2]
```diff
@@ -484,6 +484,48 @@ public struct Query<T>: ParseTypeable where T: ParseObject {
...
+                                 (excludeKeys?.sorted(by: { $0 < $1 }) ?? []))
+        let sortedFieldsKeys = ((fields?.count == 0 ? [] : ["fields"]) +
+                                (fields?.sorted(by: { $0 < $1 }) ?? []))
+
+        let sortedSets = (
+            sortedKeys +
+            sortedInclude +
+            sortedExcludeKeys +
+            sortedFieldsKeys
+        ).joined(separator: "")
```
Comment (cbaker6) [inline Sources/ParseSwift/Types/Query.swift]: When combining, it’s usually appropriate to add a delimiter such as a comma or semicolon

**Review by cbaker6 — CHANGES_REQUESTED** [review]

**Comment by cbaker6** [comment]
My view is still the same as here: https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370532155

**Comment by vdkdamian** [comment]
> The query is equatable, this shouldn’t matter. Are you saying 2 queries with the same sets don’t return equal?

Yes, I changed the test to this, and ran it 5 times.
Below you can see my test.
If you look at the image, you can see that the sets switch position randomly.
Scherm­afbeelding 2023-01-05 om 21 50 21

```
func testIdentifier() {
        let query = GameScore.query()
            .order([.ascending("points"), .descending("oldScore")])
            .exclude("hello", "world")
            .include("foo", "bar")
            .select("yolo", "nolo")
            .limit(10)

        let query2 = GameScore.query()
            .order([.ascending("points"), .descending("oldScore")])
            .exclude("hello", "world")
            .include("foo", "bar")
            .select("yolo", "nolo")
            .limit(10)

        let query3 = GameScore.query()
            .order([.ascending("teams"), .descending("points")])
            .include("yolo", "nolo")
            .limit(100)

        XCTAssertFalse(query.id.isEmpty)
        XCTAssertTrue(query.id.contains("orderpointsoldScore"))
        XCTAssertTrue(query.id.contains("excludeKeyshelloworld"))
        XCTAssertTrue(query.id.contains("includebarfoo"))
        XCTAssertTrue(query.id.contains("yolo"))
        XCTAssertTrue(query.id.contains("nolo"))
        XCTAssertTrue(query.id.contains("limit10"))
        XCTAssertEqual(query.id, query2.id)
        XCTAssertNotEqual(query2.id, query3.id)

        if let jsonData = try? ParseCoding.jsonEncoder().encode(query),
              let descriptionString = String(data: jsonData, encoding: .utf8) {
            print("descriptionString")
            print(descriptionString)
        }
        if let jsonData2 = try? ParseCoding.jsonEncoder().encode(query),
              let descriptionString2 = String(data: jsonData2, encoding: .utf8) {
            print("descriptionString2")
            print(descriptionString2)
        }
        if let jsonData3 = try? ParseCoding.jsonEncoder().encode(query),
              let descriptionString3 = String(data: jsonData3, encoding: .utf8) {
            print("descriptionString3")
            print(descriptionString3)
        }
    }
```


> I’m still not convinced a query needs to be identifiable or persisted. Even if someone needs to store it, it can be stored in JSON form and decoded to check for equality. It sounds like you would like to treat a query as a ParseObject, in which it is not. Making the properties public opens up the possibility for headaches.

What about my solution to just create a seperate variable with the code from the id?
This way it doesn't need to conform to `Identifiable`.

**Comment by cbaker6** [comment]
> What about my solution to just create a seperate variable with the code from the id?
> This way it doesn't need to conform to Identifiable.

This doesn't change anything, you might as well be identifiable then. You have not given me a good enough reason to add `id` or a variable of another name to `Query`. Your reasoning before was to persist `Query`, I don't see a reason why something like that should be supported by the SDK.

**Comment by vdkdamian** [comment]
> > What about my solution to just create a seperate variable with the code from the id?
> > This way it doesn't need to conform to Identifiable.
> 
> This doesn't change anything, you might as well be identifiable then. You have not given me a good enough reason to add `id` or a variable of another name to `Query`. Your reasoning before was to persist `Query`, I don't see a reason why something like that should be supported by the SDK.

OK

**Comment by vdkdamian** [comment]
I don't know why you think you can know why anyone would need this. I need it and it's impossible to do it on my own if the sets are changing randomly. I also can't access the internal variables but please sir tell me how to do it then.

Your sympathy has changed since I tried to solve the offline mode on the other repo. 
You are literally blocking me since I need to have a way to identify the query.

**Comment by cbaker6** [comment]
> I don't know why you think you can know why anyone would need this. I need it and it's impossible to do it on my own if the sets are changing randomly. I also can't access the internal variables but please sir tell me how to do it then.

I explain this here: https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370266839, https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370299985, and https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370532155. You have not provided compelling reasons to counter any of the points I've made. You simply just want it added. A `Query` is not a `ParseObject` and shouldn't be treated as one.

> Your sympathy has changed since I tried to solve the offline mode on the other repo.
> You are literally blocking me since I need to have a way to identify the query.

This isn't true. The idea doesn't fit the SDK, it has nothing to do with anything you choose to do with another repo.

**Comment by vdkdamian** [comment]
> I explain this here: [#42 (comment)](https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370266839), [#42 (comment)](https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370299985), and [#42 (comment)](https://github.com/netreconlab/Parse-Swift/pull/42#issuecomment-1370532155). You have not provided compelling reasons to counter any of the points I've made. You simply just want it added. A `Query` is not a `ParseObject` and shouldn't be treated as one.

I am not using it as a ParseObject, I just need it cause I am using my own way of storing objects and managing them offline.
But as mentioned, since the sets are changing, I can't identify them.
I can understand that maybe not everyone needs this. But can you maybe show me a workaround then? I really tried to find a solution without creating this PR.

> This isn't true. The idea doesn't fit the SDK, it has nothing to do with anything you choose to do with another repo.

Ok good to hear. Apologies for my assumption then.

**Comment by cbaker6** [comment]
The possible way is using the JSON to make the identifiable query you want. You can copy `Query` from the SDK and make your own version and decode the JSON in your copied version.

**Comment by vdkdamian** [comment]
> The possible way is using the JSON to make the identifiable query you want. You can copy `Query` from the SDK and make your own version and decode the JSON in your copied version.

**Comment by cbaker6** [comment]
> I need it and it's impossible to do it on my own if the sets are changing randomly.

The latest changes to the SDK makes encoding the properties that are sets deterministic, meaning they will always be in order.

**Comment by vdkdamian** [comment]
> The latest changes to the SDK makes encoding the properties that are sets deterministic, meaning they will always be in order.

Thank you! Will make my extension a lot cleaner

**Pull request closed without merging**